src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -122,26 +125,25 @@ public void onMessage(String message) {
                 @Override
                 public void onClose(int code, String reason, boolean remote) {
                     log.info("Connection closed! Code: " + code + ", Reason: " + reason);
-                    state = State.Disconnected;
                     if (remote) {
                         sleepThenReconnect();
                     } else {
+                        state = State.Disconnected;
                         StreamrClient.this.onClose();
+
+                        if (errorWhileConnecting != null) {
+                            throw new RuntimeException(errorWhileConnecting);
+                        }
                     }
                 }
 
                 @Override
                 public void onError(Exception ex) {
                     log.error(ex);
-                    if (ex instanceof IOException) {
-                        sleepThenReconnect();
-                    } else {
-                        if (state == State.Connecting) {
-                            errorWhileConnecting = ex;
-                        }
-
-                        StreamrClient.this.onError(ex);
+                    if (state == State.Connecting) {
```

Comment:
If an error causes the connection to close, both `onError` and `onClose` will be invoked. Thus before two simultaneous reconnect attempts were possible.